### PR TITLE
extmod/uzlib: Add gzip compression support.

### PR DESCRIPTION
**Pull request by andrewleech**
Exposes basic compression support from uzlib. I originally wrote this nearly a year ago so don't really remember too much about it.

Pushed up to support #5590

Doesn't have any unit tests written... probably needs some first. Possibly some #define's to enable/disable the compress functionality?

### CODE REVIEW

File: extmod/moduzlib.c [hunk 1]
```diff
@@ -206,9 +207,62 @@ STATIC mp_obj_t mod_uzlib_decompress(size_t n_args, const mp_obj_t *args) {
...
+    mp_obj_t data = args[0];
+    mp_buffer_info_t bufinfo;
+    mp_get_buffer_raise(data, &bufinfo, MP_BUFFER_READ);
+    int len = bufinfo.len;
+
+    struct uzlib_comp *comp = m_new_obj(struct uzlib_comp);
+    memset(comp, 0, sizeof(*comp));
+
+    comp->dict_size = 32768;
+    comp->hash_bits = 12;
```
Comment (stinos) [inline extmod/moduzlib.c]: This and previsou line could use a comment stating why this is chosen I think (I know nothing about gzip though)

File: extmod/moduzlib.c [hunk 2]
```diff
@@ -206,9 +207,62 @@ STATIC mp_obj_t mod_uzlib_decompress(size_t n_args, const mp_obj_t *args) {
...
+    size_t hash_size = sizeof(uzlib_hash_entry_t) * (1 << comp->hash_bits);
+    comp->hash_table = gc_alloc(hash_size, false);
+    memset(comp->hash_table, 0, hash_size);
+
+    zlib_start_block(&comp->out);
+    uzlib_compress(comp, bufinfo.buf, len);
+    zlib_finish_block(&comp->out);
+
+    printf("compressed from %u to %u raw bytes\n", len, comp->out.outlen);
```
Comment (stinos) [inline extmod/moduzlib.c]: Use `DEBUG_printf` ?

File: extmod/moduzlib.c [hunk 3]
```diff
@@ -206,9 +207,62 @@ STATIC mp_obj_t mod_uzlib_decompress(size_t n_args, const mp_obj_t *args) {
...
+
+    mp_uint_t dest_buf_size = (comp->out.outlen + 6 + (3*sizeof(int)));
```
Comment (stinos) [inline extmod/moduzlib.c]: Same remark as earlier, what are these magic constants?

**Comment by stinos** [comment]
There's MICROPY_PY_UZLIB already, but I guess splitting that up in separate compress/decompress doesn't hurt and can be used to keep the original behavior (i.e. not enabling compression by default).

A bunch of tests would be nice indeed, and fairly easy to write (assuming asserting that decompress(compress(x)) == x is sufficient for proving it works)

File: extmod/moduzlib.c (hunk 1 above)
Comment (codefreax) [reply extmod/moduzlib.c]: Would is make sense to expose these settings in the API?

File: extmod/moduzlib.c [hunk 4]
```diff
@@ -206,9 +207,62 @@ STATIC mp_obj_t mod_uzlib_decompress(size_t n_args, const mp_obj_t *args) {
...
+    byte *dest_buf = m_new(byte, dest_buf_size);
+
+    int i = 0;
+    dest_buf[i++] = 0x1f;
+    dest_buf[i++] = 0x8b;
+    dest_buf[i++] = 0x08;
+    dest_buf[i++] = 0x00; // FLG
+    int mtime = 0;
+    memcpy(&dest_buf[i], &mtime, sizeof(mtime));
+    i += sizeof(mtime);
```
Comment (codefreax) [inline extmod/moduzlib.c]: According to [Wikipedia](https://en.wikipedia.org/wiki/Gzip), `mtime` must be 4 bytes. Use `uint32_t` here?

File: extmod/moduzlib.c [hunk 5]
```diff
@@ -206,9 +207,62 @@ STATIC mp_obj_t mod_uzlib_decompress(size_t n_args, const mp_obj_t *args) {
...
 
 #if !MICROPY_ENABLE_DYNRUNTIME
+STATIC mp_obj_t mod_uzlib_gzip(size_t n_args, const mp_obj_t *args) {
+    mp_obj_t data = args[0];
+    mp_buffer_info_t bufinfo;
+    mp_get_buffer_raise(data, &bufinfo, MP_BUFFER_READ);
+    int len = bufinfo.len;
+
+    struct uzlib_comp *comp = m_new_obj(struct uzlib_comp);
+    memset(comp, 0, sizeof(*comp));
```
Comment (codefreax) [inline extmod/moduzlib.c]: nit: Merge into `m_new0(struct uzlib_comp, 1);`?

File: extmod/moduzlib.c [hunk 6]
```diff
@@ -206,9 +207,62 @@ STATIC mp_obj_t mod_uzlib_decompress(size_t n_args, const mp_obj_t *args) {
...
+    printf("compressed from %u to %u raw bytes\n", len, comp->out.outlen);
+
+    mp_uint_t dest_buf_size = (comp->out.outlen + 6 + (3*sizeof(int)));
+    byte *dest_buf = m_new(byte, dest_buf_size);
+
+    int i = 0;
+    dest_buf[i++] = 0x1f;
+    dest_buf[i++] = 0x8b;
+    dest_buf[i++] = 0x08;
+    dest_buf[i++] = 0x00; // FLG
```
Comment (codefreax) [inline extmod/moduzlib.c]: nit: Can you add a few more comments here and below where there are more magic numbers?

File: extmod/moduzlib.c [hunk 7]
```diff
@@ -206,9 +207,62 @@ STATIC mp_obj_t mod_uzlib_decompress(size_t n_args, const mp_obj_t *args) {
...
+    i += sizeof(mtime);
+    dest_buf[i++] = 0x04; // XFL
+    dest_buf[i++] = 0x03; // OS
+
+    memcpy(&dest_buf[i], comp->out.outbuf, comp->out.outlen);
+    i += comp->out.outlen;
+
+    unsigned int crc = ~uzlib_crc32(bufinfo.buf, len, ~0);
+    memcpy(&dest_buf[i], &crc, sizeof(crc));
+    i += sizeof(crc);
```
Comment (codefreax) [inline extmod/moduzlib.c]: `uint32_t`?

File: extmod/moduzlib.c [hunk 8]
```diff
@@ -206,9 +207,62 @@ STATIC mp_obj_t mod_uzlib_decompress(size_t n_args, const mp_obj_t *args) {
 STATIC MP_DEFINE_CONST_FUN_OBJ_VAR_BETWEEN(mod_uzlib_decompress_obj, 1, 3, mod_uzlib_decompress);
 
 #if !MICROPY_ENABLE_DYNRUNTIME
+STATIC mp_obj_t mod_uzlib_gzip(size_t n_args, const mp_obj_t *args) {
+    mp_obj_t data = args[0];
+    mp_buffer_info_t bufinfo;
+    mp_get_buffer_raise(data, &bufinfo, MP_BUFFER_READ);
+    int len = bufinfo.len;
```
Comment (codefreax) [inline extmod/moduzlib.c]: This is written to the output; should this be `uint32_t` (or whatever the right size is)?

**Comment by QAMU** [comment]
hello,
for this commit, how to compress data with micropython.
can I use uzlib.compress(data)?

**Comment by andrewleech** [comment]
Hi @QAMU I think it worked like `uzlib.gzip(data)`
I haven't used it for some time though, not sure it's ever likely to get completed enough to merge.

**Comment by QAMU** [comment]
Hi @andrewleech,
thanks for your reply, can I use uzlib.decompress() to decompress data = uzlib.gzip()?

**Comment by andrewleech** [comment]
yep the two should work together well

**Comment by QAMU** [comment]
yes, but uzlib.decompress() is work just for zlib compression not gzip?
so I can't decompress my data compressed by gzlib.
thank you Sir

**Comment by harbaum** [comment]
It seems compression and decompression don't work with each other:

```
>>> uzlib.decompress(uzlib.gzip("Hello World!"))
compressed from 12 to 14 raw bytes
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
ValueError: -3
```

**Comment by harbaum** [comment]
> It seems compression and decompression don't work with each other:
They do if one removes the gzip header, crc and length fields and uses the correct wbits:

```
>>> uzlib.decompress(uzlib.gzip("Hello World!")[10:-8], -15)
compressed from 12 to 14 raw bytes
bytearray(b'Hello World!')
```

**Comment by harbaum** [comment]
Updated attempt [here](https://github.com/micropython/micropython/pull/6972).

**Comment by QAMU** [comment]
to compress use: 
```
import uzlib 
def compress(buffer):
    encoded = = uzlib.gzip(buffer)
    return encoded
```
to decompress:

```
import uzlib 
import uio
def decompress(buffer):
    file = uio.BytesIO(buffer)
    decoded = uzlib.DecompIO(file, 31)
    return decoded.read()
```

**Review by harbaum — COMMENTED** [review]
> to decompress:
> 
> ```
> import uzlib 
> import uio
> def decompress(buffer):
>     file = uio.BytesIO(buffer)
>     decoded = uzlib.DecompIO(file, 31)
>     return decoded.read()
> ```

Why not:

```
def decompress(buffer):
  return uzlib.decompress(buffer[10:-8], -15)

```
Seems more lightweight for an embedded solution.

File: extmod/moduzlib.c [hunk 9]
```diff
@@ -206,9 +207,62 @@ STATIC mp_obj_t mod_uzlib_decompress(size_t n_args, const mp_obj_t *args) {
...
+    memset(comp, 0, sizeof(*comp));
+
+    comp->dict_size = 32768;
+    comp->hash_bits = 12;
+    size_t hash_size = sizeof(uzlib_hash_entry_t) * (1 << comp->hash_bits);
+    comp->hash_table = gc_alloc(hash_size, false);
+    memset(comp->hash_table, 0, hash_size);
+
+    zlib_start_block(&comp->out);
+    uzlib_compress(comp, bufinfo.buf, len);
```
[the remark this answers is not in the document]
Comment (kindmartin) [reply extmod/moduzlib.c:226]: > `uzlib_compress` as well as `zlib_start_block` internally (re-)allocate a buffer at `comp->out` which is never free'd. As a result the system runs out of memory after a few compression runs.
> 
> A `free(comp->out.outbuf);` after the subsequent memcpy solved this.

Im trying to encode a string to get into the shorter sms data possible., 
questio is : how can I add this gzip uzlib extension to my current 1.14 toolchain for esp32?

**Comment by andrewleech** [comment]
@harbaum Thanks for your additions in #6972, there is some great work there. I've rebased your changes on my original branch as a separate commit to keep your attribution.

In addition, I've added support for gzip header in the `decompress(bytes, 31)` function to allow it to work directly on the gzip compressed data.

The docs have been updated and I've added a basic compress unit test.

**Comment by andrewleech** [comment]
Ok I've broken the unit tests with the last additions of gzip decompress and rebase onto current master... and it's too big for tiny ports - so probably does need a new feature flag. Either that, or just build & distribute in the dynamic C module version of the uzlib.

**Comment by dpgeorge** [comment]
Superseded by #11905.

**Pull request closed without merging**